core/src/main/java/tachyon/worker/rdma/RDMADataServer.java

```diff
@@ -0,0 +1,241 @@
+      }
+      SessionServerCallbacks callbacks = new SessionServerCallbacks(uri);
+      ServerSession session = new ServerSession(sesKey, callbacks);
+      callbacks.setSession(session);
+      mListener.accept(session);
+    }
+  }
+
+  private final class SessionServerCallbacks implements ServerSession.Callbacks {
+    private final DataServerMessage mResponseMessage;
+    private ServerSession mSession;
+
+    public SessionServerCallbacks(URI uri) {
+      List<NameValuePair> params = URLEncodedUtils.parse(uri, "UTF-8");
+      long blockId = Long.parseLong(params.get(0).getValue());
+      long offset = Long.parseLong(params.get(1).getValue());
+      long length = Long.parseLong(params.get(2).getValue());
+      LOG.debug("got request for block id {} with offset {} and length {}", blockId, offset,
+          length);
+      final int lockId = mBlocksLocker.getLockId();
+      final StorageDir storageDir = mBlocksLocker.lock(blockId, lockId);
+      ByteBuffer data;
+      int dataLen = 0;
+      try {
+        data = storageDir.getBlockData(blockId, offset, (int) length);
+        storageDir.accessBlock(blockId);
+        dataLen = data.limit();
+      } catch (Exception e) {
+        LOG.error(e.getMessage(), e);
+        data = null;
+      }
+      mResponseMessage =
+          DataServerMessage.createBlockResponseMessage(true, blockId, offset, dataLen, data);
+      mResponseMessage.setLockId(lockId);
+    }
+
+    public void setSession(ServerSession ses) {
+      mSession = ses;
+    }
+
+    public void onRequest(Msg m) {
+      if (mSession.getIsClosing()) {
+        mSession.discardRequest(m);
+      } else {
+        mResponseMessage.copyMsgToBuffer(m.getOut());
+        try {
+          mSession.sendResponse(m);
+        } catch (JxioGeneralException e) {
+          LOG.error("Exception accured while sending message {}", e.toString());
+          mSession.discardRequest(m);
+        } catch (JxioSessionClosedException e) {
+          LOG.error("session was closed unexpectedly {}", e.toString());
+          mSession.discardRequest(m);
+        } finally {
+          if (mResponseMessage.finishSending()) {
+            mSession.close();
+          }
+        }
+      }
+    }
+
+    public void onSessionEvent(EventName session_event, EventReason reason) {
+      LOG.debug("got event {}, the reason is {}", session_event.toString(), reason.toString());
+      if (session_event == EventName.SESSION_CLOSED) {
+        mResponseMessage.close();
+        mBlocksLocker.unlock(Math.abs(mResponseMessage.getBlockId()), mResponseMessage.getLockId());
+      }
+    }
+
+    public boolean onMsgError(Msg msg, EventReason reason) {
+      LOG.error("{} onMsgErrorCallback. reason is {}", this.toString(), reason);
+      return true;
+    }
+  }
+
+  @Override
+  public void run() {
+    int ret =
+        mEventQueueHandler.runEventLoop(EventQueueHandler.INFINITE_EVENTS,
+            EventQueueHandler.INFINITE_DURATION);
+    if (ret == -1) {
```

Comment:
A failure here is catastrophic and a system restart should be done to start the server again?